master/pkg/searcher/asha_stopping.go

```diff
@@ -0,0 +1,255 @@
+			TrialRungs:       make(map[model.RequestID]int),
+			EarlyExitTrials:  make(map[model.RequestID]bool),
+			ClosedTrials:     make(map[model.RequestID]bool),
+			SearchMethodType: ASHASearch,
+		},
+	}
+}
+
+func (s *asyncHalvingStoppingSearch) Snapshot() (json.RawMessage, error) {
+	return json.Marshal(s.asyncHalvingSearchState)
+}
+
+func (s *asyncHalvingStoppingSearch) Restore(state json.RawMessage) error {
+	return json.Unmarshal(state, &s.asyncHalvingSearchState)
+}
+
+// promotions handles bookkeeping of validation metrics and decides whether to continue
+// training the current trial.
+func (r *rung) continueTraining(requestID model.RequestID, metric float64, divisor float64) bool {
+	// Compute cutoff for promotion to next rung to continue training.
+	numPromote := max(int(float64(len(r.Metrics)+1)/divisor), 1)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.Metrics),
+		func(i int) bool { return r.Metrics[i].Metric >= metric },
+	)
+	// We will continue training if trial ranked in top 1/divisor for the rung or
+	// if there are fewere than divisor trials in the rung.
+	promoteNow := insertIndex < numPromote
+
+	r.Metrics = append(r.Metrics, trialMetric{})
+	copy(r.Metrics[insertIndex+1:], r.Metrics[insertIndex:])
+	r.Metrics[insertIndex] = trialMetric{
+		RequestID: requestID,
+		Metric:    metric,
+		Promoted:  promoteNow,
+	}
+
+	return promoteNow
+}
+
+func (s *asyncHalvingStoppingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// The number of initialOperations will control the degree of parallelism
+	// of the search experiment since we guarantee that each validationComplete
+	// call will return a new train workload until we reach MaxTrials.
+
+	// We will use searcher config field if available.
+	// Otherwise we will default to a number of trials that will
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials > 0 {
+		maxConcurrentTrials = min(s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		s.TrialRungs[create.RequestID] = 0
+		ops = append(ops, create)
+		ops = append(ops, NewTrain(create.RequestID, s.Rungs[0].UnitsNeeded))
+		ops = append(ops, NewValidate(create.RequestID))
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingStoppingSearch) trialCreated(
+	ctx context, requestID model.RequestID) ([]Operation, error) {
+	s.Rungs[0].OutstandingTrials++
+	s.TrialRungs[requestID] = 0
+	return nil, nil
+}
+
+func (s *asyncHalvingStoppingSearch) trialClosed(
+	ctx context, requestID model.RequestID) ([]Operation, error) {
+	s.TrialsCompleted++
+	s.ClosedTrials[requestID] = true
+	return nil, nil
+}
```

Comment:
thinking to the future a bit, does asha (or any searcher) really need to be aware of a trial _actually_ starting or stopping, or could this happen with newcreate/newclose. trying to think of ways to minimize the searcher interface as we move towards sprinkle api?

Reply:
on the same note, how do you feel about removing train operations (or at the the callback to expect them back, i dont think anything meaningfully hooks into training completed except maybe progress or something)

Reply:
i guess as a more directed question, are there any hp search algorithms that need to know training metrics, when the scheduler started trials, or when the trial responded to a close and terminated?

Reply:
agree that training metrics aren't really relevant for HP search.  

there are definitely hp algos that need to track how many hps have been evaluated/are being evaluated and what each step the validation metrics correspond to.  in the case of asha for example, the algo presumably has a max_trials so it needs to know how many trials total have been created in the bottom rung, including ones that are in progress and do not have a validation metric associated with it.  similarly, for PBT, the algorithm needs to know how many trials have been launched for a particular evolution round.  

in a sprinkle world, hp workers would go to our hp searcher and say give me a workload and once it finishes, it returns the validation metric associated to the searcher and requests the next workload.  the searcher will be in charge of telling a worker there are no more workloads remaining and to exit.  being able to detect the stopping condition would likely require knowing when a trial is created.  in this scheme, the workers will also need to tell the searchers which trials to exclude from further consideration (e.g. to handle invalidhp or early exit trials).